### PR TITLE
Advent 2016: Data Pipelines and Versioning with the Pachyderm Go Client

### PR DESCRIPTION
**Pull request by dwhitena**
Ok, here's my data post (what else would you expect from me, haha).  It's fairly long, but I tried to give a complete walkthrough of data pipelining via the Pachyderm Go client.  Let me know what questions, suggestions, etc. you have.  

btw, really enjoying reading the articles.  Great work on organizing this and really enjoying seeing the quality posts by all the authors.

### CODE REVIEW

**Comment by dwhitena** [comment]
Going to shorten the longer code snippets this weekend to make things a little more readable.

**Comment by dgryski** [comment]
\o/ https://github.com/campoy/embedmd

File: content/advent-2016/pachyderm.md [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+
+As mentioned, you can build your Pachyderm data pipelines using any languages or frameworks (Go, python, Tensorflow, Spark, Rust, etc.), but Pachyderm does have a nice [Go client](https://godoc.org/github.com/pachyderm/pachyderm/src/client) that will let you launch pipelines, put data into data versioning, pull data out of data versioning, etc. directly from your Go applications.  For example, you could commit metrics from your Go backend service directly to Pachyderm and, on every commit, have Pachyderm automatically update predictive analysis (written in Go, Tensorflow, python, or whatever you might prefer) to detect fraudulent activity based on those metrics.
```
Comment (bradleyfalzon) [inline content/advent-2016/pachyderm.md]: I think `python` should also be `Python`

File: content/advent-2016/pachyderm.md [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+
+To read more about the Pachyderm project, visit [Pachyderm's website](http://pachyderm.io/) and look through [the docs](http://docs.pachyderm.io/en/latest/). 
+
+# A data processing example for this post
+
+In this post, we are going to illustrate some distributed data processing and data versioning with a few simple Go programs and some Pachyderm configuration. The goal of the data processing will be to generate statistics about Go projects posted to [Github](github.com).  We will:
+
+1. Deploy Pachyderm.
+
+2. Create a Pachyderm pipeline that takes Go repository names (e.g., `github.com/docker/docker`) as input and outputs as couple of stats/metrics about those repositories.
```
Comment (bradleyfalzon) [inline content/advent-2016/pachyderm.md]: Should `as couple of` be `as a couple of`?

File: content/advent-2016/pachyderm.md [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+projects            18 seconds ago      0 B                 
+stats               5 seconds ago       0 B                 
+$ pachctl list-pipeline
+NAME                INPUT               OUTPUT              STATE               
+stats               projects            stats               running 
+```
+
+Notice that an output repository (with the name of our pipeline) has also been created.  The output of our `stats` pipeline will be versioned there.  
+
+Now, let's commit some data into the input data repository called `projects`.  Specifically we will commit a first file `one.txt` into the `projects` repository on the `master` branch, where `one.txt` includes the Github repository name that we want to analyze (`docker/docker`):
```
Comment (bradleyfalzon) [inline content/advent-2016/pachyderm.md]: `Github` should be `GitHub`

File: content/advent-2016/pachyderm.md [hunk 4]
```diff
@@ -0,0 +1,519 @@
...
+d9255dbe5df83e25ba3d3c890e022598   stats/cd309fd23a3745b68dafe56b581d63bf/0   2 minutes ago       -                   running    
+d15c3e962e2ed2412be460db26a21c85   stats/46b8270e1e2c4fdd990d89eaa4b4351b/0   2 minutes ago       -                   running    
+7d7a88b806fefcf8f2166492eb7e32a6   stats/65e0d4edaca7483fa06c7c2443133774/0   2 minutes ago       -                   running    
+35f6c95d843a12f506e3d18c94dfffab   stats/ac98664623a74f029506ec10701387c1/0   2 minutes ago       -                   running    
+fd5b8623ca2556e0da4b71fc205c6f52   stats/f7ebb735f2fa46b894d3ce81e8cb8855/0   2 minutes ago       -                   running    
+16f1108c71e6a0da6dca4712e54d4ddb   stats/ff7a262a03ea4f279e92c6d54c7e8129/0   2 minutes ago       -                   running    
+4c4f53668e46c20fdeb1286ca971ea1f   stats/930e29b046a948649176b04225a547d9/0   2 minutes ago       -                   running   
+```
+
+Eventually these will finish, and we will see a each corresponding commit to our output data repository `stats`:
```
Comment (bradleyfalzon) [inline content/advent-2016/pachyderm.md]: Should `see a each` be `see each`?

File: content/advent-2016/pachyderm.md [hunk 5]
```diff
@@ -0,0 +1,519 @@
...
+
+As mentioned, you can build your Pachyderm data pipelines using any languages or frameworks (Go, python, Tensorflow, Spark, Rust, etc.), but Pachyderm does have a nice [Go client](https://godoc.org/github.com/pachyderm/pachyderm/src/client) that will let you launch pipelines, put data into data versioning, pull data out of data versioning, etc. directly from your Go applications.  For example, you could commit metrics from your Go backend service directly to Pachyderm and, on every commit, have Pachyderm automatically update predictive analysis (written in Go, Tensorflow, python, or whatever you might prefer) to detect fraudulent activity based on those metrics.
+
+To read more about the Pachyderm project, visit [Pachyderm's website](http://pachyderm.io/) and look through [the docs](http://docs.pachyderm.io/en/latest/). 
+
+# A data processing example for this post
+
+In this post, we are going to illustrate some distributed data processing and data versioning with a few simple Go programs and some Pachyderm configuration. The goal of the data processing will be to generate statistics about Go projects posted to [Github](github.com).  We will:
```
Comment (bradleyfalzon) [inline content/advent-2016/pachyderm.md]: `Github` should be `GitHub`

File: content/advent-2016/pachyderm.md [hunk 6]
```diff
@@ -0,0 +1,519 @@
...
+rm dep.log
+
+# get number of lines of go code
+golines=`( find $pkgPath -name '*.go' -print0 | xargs -0 cat ) | wc -l`
+
+# output the stats
+echo $REPONAME, $deps, $golines
+```
+
+This includes the `wc -l` and `go list` commands along with some clean up and things to support [Godep](https://github.com/tools/godep).  The script will output our metrics given a Github Go project name input to the variable `REPONAME`.  Note, the metrics are output to `/pfs/out`, which is a way of telling Pachyderm to output something to the output data repository of a pipeline stage (which will be explained further below).  Our Docker image is simply the `golang` image plus this script:
```
Comment (bradleyfalzon) [inline content/advent-2016/pachyderm.md]: `Github` should be `GitHub`

File: content/advent-2016/pachyderm.md [hunk 7]
```diff
@@ -0,0 +1,519 @@
...
+
+In essense, when new data is committed to a data repository call `projects`, this pipeline will be triggered and process the data using the specified image, cmd, and stdin.  The pipeline only has one "stage."  That is, we are only specifying one data processing step that takes data from the `projects` repository and outputs our metrics.  If we wanted to have more stages in our pipeline (e.g., a stage that compiles the project and measures compile time), we would just need to add more JSON blobs to `pipeline.json`.  See Pachyderm's [fruit stand example](http://pachyderm.readthedocs.io/en/stable/getting_started/beginner_tutorial.html) to see a multistage pipeline in action.
+
+## Step 2b: Run and Test the Pachyderm Pipeline
+
+At this point, we have the following:
+
+- A script called `stats.sh` that calculates metrics for a Go project given a Github repository name.
```
Comment (bradleyfalzon) [inline content/advent-2016/pachyderm.md]: `Github` should be `GitHub`

File: content/advent-2016/pachyderm.md [hunk 8]
```diff
@@ -0,0 +1,519 @@
...
+	}
+
+	// Create a repo called "projects."
+	if err := c.CreateRepo("projects"); err != nil {
+		log.Fatal(err)
+	}
+}
+```
+
+Then add a loop that commits files to `projects`, where each successive file includes the name of a different Go project on Github:
```
Comment (bradleyfalzon) [inline content/advent-2016/pachyderm.md]: `Github` should be `GitHub`

File: content/advent-2016/pachyderm.md [hunk 9]
```diff
@@ -0,0 +1,519 @@
...
+- [Join the Pachyderm Slack community](http://slack.pachyderm.io/) for questions, discussions, deployment help, etc.
+- Read [the Pachyderm docs](http://docs.pachyderm.io/en/latest/).
+- Read [the godocs](https://godoc.org/github.com/pachyderm/pachyderm/src/client) for the Pachyderm Go client.
+- Check out [example Pachyderm pipelines](http://docs.pachyderm.io/en/latest/examples/readme.html).
+- Connect with Pachyderm [on Twitter](https://twitter.com/pachydermIO).
+
+Also, more generally, make sure to get involved in the Go data science community:
+
+- Join the `#data-science` channel on [Gophers Slack](https://invite.slack.golangbridge.org/).
+- Check out the [gopherds org on Github](https://github.com/gopherds).  Specifically, visit the [resources repo](https://github.com/gopherds/resources) for [a list of Go data science packages](https://github.com/gopherds/resources/tree/master/tooling) and [community updates](https://github.com/gopherds/resources/tree/master/community).
```
Comment (bradleyfalzon) [inline content/advent-2016/pachyderm.md]: `Github` should be `GitHub`

**Comment by dwhitena** [comment]
Thank you so much @sbinet and @bradleyfalzon!  I have added your suggested changes.  Also, @dgryski, I have shortened the Go code snippets a bit.

**Pull request merged**